docs/src/userguide/freethreading.rst

```diff
@@ -160,3 +160,297 @@ to require extreme care to make it work correctly).
+threads for example by having the different threads perform completely isolated
+blocks of work which are only collected at the end.  Python code is no exception -
+especially since Python reference counting means that even apparently "read-only"
+operations usually actually involve both reading and writing.
+
+As an example consider a program that collects unique words from multiple files.
+In this case it would probably be best to read each file to a separate ``set``
+and then combine them at the end::
+
+  def read_from_files_good(filenames):
+    def read_from_file(filename):
+      out = set()
+      with open(filename, 'r') as f:
+        for line in f:
+          words = line.split()
+          for word in words:
+            out.add(word)
+      return out
+
+    overall_result = set()
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+      for file_result in executor.map(read_from_file, filenames):
+        overall_result.update(file_result)
+    return overall_result
+
+rather than updating one ``set`` from all threads::
+
+  def read_from_files_bad(filenames):
+    overall_result = set()
+    def read_from_file(filename):
+      with open(filename, 'r') as f:
+        for line in f:
+          words = line.split()
+          for word in words:
+            overall_result.add(word)
+
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+      for _ in executor.map(read_from_file, filenames):
+        pass
+    return overall_result
+
+The less your threads interact, the less chance there is for bugs, the less
+need there is for locking to control their interaction, and the less likely
+they are to slow each other down by invaliding the CPU cache for other
+threads.
+
+Should you use ``prange``?
+--------------------------
+
+Although ``prange`` is the parallelization mechanism *built in* to Cython, it
+is not the only option, and probably should not be your default option.
+
+``prange`` is a fairly thin wrapper over OpenMP's "parallel for".  This means
+it is ideal for problems where you have a big loop, every iteration is basically
+the same, and the result of each iteration is independent of any other iteration.
+If this does not describe your problem then ``prange`` is probably not the
+solution.
+
+Remember that all the threading options available in Python are also available
+in Cython.  For example you can start threads with ``threading.Thread`` or
+``concurrent.futures.ThreadPoolExecutor`` and this is much more flexible than
+``prange``.  Similarly the synchronization tools in ``threading.Thread``
+are also available in Cython.
+
+Try to avoid Python code in ``prange``
+--------------------------------------
+
+``prange`` has some slightly unintuitive behaviour about which data is
+shared and which isn't.  Typically C variables (e.g. ``int``, ``float``) are
+treated as "thread-local" and so each thread has its own copy. However,
+Python object variables are treated as shared between all the threads.
```

Comment:
`int` and `float` do not seem the optimal pair of examples for non-Python types.
```suggestion
``prange`` has some slightly unintuitive behaviour about which data is
shared and which isn't.  Typically C variables (e.g. ``int``, ``double``) are
treated as "thread-local" and so each thread has its own copy. However,
Python object variables are treated as shared between all the threads.
```